srml/authority-discovery/src/lib.rs

```diff
@@ -55,79 +46,40 @@ decl_module! {
 }
 
 impl<T: Trait> Module<T> {
-	/// Returns own authority identifier iff it is part of the current authority
-	/// set, otherwise this function returns None. The restriction might be
-	/// softened in the future in case a consumer needs to learn own authority
-	/// identifier.
-	fn authority_id() -> Option<T::AuthorityId> {
-		let authorities = Keys::<T>::get();
-
-		let local_keys = T::AuthorityId::all();
-
-		authorities.into_iter().find_map(|authority| {
-			if local_keys.contains(&authority) {
-				Some(authority)
-			} else {
-				None
-			}
-		})
-	}
-
 	/// Retrieve authority identifiers of the current authority set.
-	pub fn authorities() -> Vec<T::AuthorityId> {
-		Keys::<T>::get()
-	}
-
-	/// Sign the given payload with the private key corresponding to the given authority id.
-	pub fn sign(
-		payload: &Vec<u8>,
-	) -> Option<(
-		<<T as Trait>::AuthorityId as RuntimeAppPublic>::Signature,
-		T::AuthorityId,
-	)> {
-		let authority_id = Module::<T>::authority_id()?;
-		authority_id.sign(payload).map(|s| (s, authority_id))
+	pub fn authorities() -> Vec<AuthorityId> {
+		Keys::get()
 	}
 
-	/// Verify the given signature for the given payload with the given
-	/// authority identifier.
-	pub fn verify(
-		payload: &Vec<u8>,
-		signature: <<T as Trait>::AuthorityId as RuntimeAppPublic>::Signature,
-		authority_id: T::AuthorityId,
-	) -> bool {
-		authority_id.verify(payload, &signature)
-	}
-
-	fn initialize_keys(keys: &[T::AuthorityId]) {
+	fn initialize_keys(keys: &[AuthorityId]) {
 		if !keys.is_empty() {
-			assert!(Keys::<T>::get().is_empty(), "Keys are already initialized!");
-			Keys::<T>::put(keys);
+			assert!(Keys::get().is_empty(), "Keys are already initialized!");
```

Comment:
```suggestion
			assert!(Keys::exists(), "Keys are already initialized!");
```

Does `Keys::exists()` work for your needs here?

Reply:
So far I have tried to be consistent with https://github.com/paritytech/substrate/blob/master/srml/im-online/src/lib.rs#L506. In case we want to use `exists` I am happy to do that, but rather as a follow up in both modules at once.

Reply:
`exits()` is not correct here. We could use `decode_len()`, but that makes no real difference here, as this is only called at genesis. `Keys` will always be initialized by the `GenesisConfig`, but we need to make sure that we did not set them to something before. So, we check if the set keys are empty.